camel-core/src/main/java/org/apache/camel/builder/FluentProducerTemplate.java

```diff
@@ -0,0 +1,350 @@
+     */
+    public FluentProducerTemplate withExchange(final Exchange exchange) {
+        return withExchange(() -> exchange);
+    }
+
+    /**
+     * Set the exchangeSupplier which will be invoke to get the exchange to be
+     * used for send.
+     *
+     * @param exchangeSupplier
+     * @return
+     */
+    public FluentProducerTemplate withExchange(final Supplier<Exchange> exchangeSupplier) {
+        this.exchangeSupplier = exchangeSupplier;
+        return this;
+    }
+
+    /**
+     * Set the processor to use for send/request.
+     *
+     * <pre>
+     * {@code
+     * FluentProducerTemplate.on(context)
+     *     .withProcessor(
+     *         exchange -> {
+     *             exchange.getIn().setHeader("Key1", "Val1")
+     *             exchange.getIn().setHeader("Key2", "Val2")
+     *             exchange.getIn().setBody("the body")
+     *         }
+     *      )
+     *     .to("direct:start")
+     *     .request()
+     * </pre>
+     *
+     * @param processor
+     * @return
+     */
+    public FluentProducerTemplate withProcessor(final Processor processor) {
+        return withProcessor(() -> processor);
+    }
+
+    /**
+     * Set the processorSupplier which will be invoke to get the processor to be
+     * used for send/request.
+     *
+     * @param processorSupplier
+     * @return
+     */
+    public FluentProducerTemplate withProcessor(final Supplier<Processor> processorSupplier) {
+        this.processorSupplier = processorSupplier;
+        return this;
+    }
+
+    /**
+     * Set the message body
+     *
+     * @param endpointUri the endpoint URI to send to
+     * @return ProducerTemplate builder
+     */
+    public FluentProducerTemplate to(String endpointUri) {
+        return to(context.getEndpoint(endpointUri));
+    }
+
+    /**
+     * Set the message body
+     *
+     * @param endpoint the endpoint to send to
+     * @return ProducerTemplate builder
+     */
+    public FluentProducerTemplate to(Endpoint endpoint) {
+        this.endpoint = endpoint;
+        return this;
+    }
+
+    // ************************
+    // REQUEST
+    // ************************
+
+    /**
+     * Send to an endpoint returning any result output body.
+     *
+     * @return the result
+     * @throws CamelExecutionException
+     */
+    public Object request() throws CamelExecutionException {
+        return request(Object.class);
+    }
+
+    /**
+     * Send to an endpoint.
+     *
+     * @param type the expected response type
+     * @return the result
+     * @throws CamelExecutionException
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T request(Class<T> type) throws CamelExecutionException {
+        validate();
+
+        T result;
+        if (type == Exchange.class) {
+            result = (T)template().request(endpoint, processorSupplier.get());
+        } else if (type == Message.class) {
+            result = (T)template().request(endpoint, processorSupplier.get()).getIn();
```

Comment:
Ah here you may need to return OUT instead. eg do a hasOut ? getOut : getIn code on the exchange, so the correct Message instance is returned.